### ISSUE TITLE
udelim should expose a way to annotate with syntax properties instead of #%xxx

### ISSUE THREAD

**Issue by SuzanneSoy** [willghatch/racket-udelim#2]
It would be nice to be able to read `⟅a b c⟆` as `(a b c)`, but with a user-specified syntax property, instead of reading it as `(#%s-shaped-bag-delim a b c)`.

**Comment by jackfirth** [comment]
The [`'paren-shape` property](http://docs.racket-lang.org/reference/reader.html?q=paren-shape#%28idx._%28gentag._30._%28lib._scribblings%2Freference%2Freference..scrbl%29%29%29) is also an option, particularly when combined with an `#%app` macro that looks at paren shape.

**Comment by SuzanneSoy** [comment]
@jackfirth I thought about this, but won't this cause compatibility problems? For now, `'paren-shape` can only contain two distinct characters, `#\[` and `#\{`. For these custom parentheses,
* the existing tools won't expect to see any other character, which may cause errors
* it would be desirable to store two characters, "⟅⟆", possibly three if udelim at some point supports alternatives to the ".", e.g `⦃a b ⦂ c⦄` could read as `(a b . c)` with a `'paren-shape` syntax-property of "⦃⦂⦄". Although `#%dot` with a `'dot-shape` on it would probably be a better fit for this use-case.

If putting unusual values in `'paren-shape` does not risk causing any issues with existing tools, then by all means it should be used, and a convention to indicate which were the left and right parens (and possibly the dot) should be adopted, so that tools like `scribble` can reconstruct the original parentheses when rendering code samples, without having to hard-code that `⟅` gets closed by `⟆`, and so on (which would be needed if only the first character was stored, as is the case currently for `#\[` and `#\{`).

**Comment by willghatch** [comment]
Adding user-specified syntax properties would be nice.  I was going to add `'paren-shape` at one point, but it has problems as you noted, especially the fact that it only stores one of the delimiters.  Most delimiters that people would actually use have unicode properties linking the left and right parens, so that might be something that could be worked around.  But not even all of the characters I have used for delimiters have the unicode property, and there is nothing stopping people from using arbitrary characters as left and right delimiters.  Also, at some point I would like to support multi-character delimiters (probably in an incompatible udelim2, though, due to readtable issues) which would cause even more issues with the 'paren-shape property.

User-specified syntax properties would be easy to support though.  I could just add keyword arguments to `make-list-delim-readtable` -- maybe `#:syntax-property` and `#:syntax-property-value` that both default to `#f`.  In fact, maybe the `/wrap` versions really should have been keyword arguments as well.  What do you guys think?

**Comment by SuzanneSoy** [comment]
Sounds like a reasonable course of action to me.

**Comment by SuzanneSoy** [comment]
Good point concerning multi-character delimiters, I guess the "extended" version of `'paren-shape`, if it makes sense compatibility-wise, should then be a list of two strings, `(list "opening" "closing")` (and possibly three if changing the dot `.` makes sense) or something similar, instead of just being a string with two characters as I suggested.

**Comment by willghatch** [comment]
FYI as to why I haven't done anything about this yet, I'm looking into some deeper reader things, and I don't want to commit to something with respect to this issue until I do that.
